### ISSUE TITLE
problem with short burst and random pattern

### ISSUE THREAD

**Issue by AnttiLukats** [jorisvr/te0890-utils#1]
when using burst len 1 and 2 and random pattern we see errors in every round, they are about the same amount of errors at random locations.

when the burst length array is modified from 1,2,3.. to 3,3,3.. there are no more errors, long time looped test work without errors.

current test length 28+ hours

**Comment by jorisvr** [comment]
To add some context in case someone else reads this issue later:
Your message refers to the HyperRAM memory test design, ported to a different Trenz Spartan-7 board.
https://shop.trenz-electronic.de/de/CR00107-01-CRUVI-carrier-board-with-AMD-Spartan-7

Same issue also discussed here:
https://forum.trenz-electronic.de/index.php/topic,1320.msg8188.html#msg8188

I noticed your comment in https://github.com/MJoergen/HyperRAM/issues/2
It looks like that potentially explains all problems you have seen.

There is unfortunately not much I can do about it. I don't have a RevD RAM chip and I don't want to change the VHDL code without testing. I will update the README file to warn that my HyperRAM controller is not work reliable with currently available devices.

**Comment by AnttiLukats** [comment]
CONFIRMED this bug is the rev D errata, you must issue 2 full clock write sequence with all writes, shorter writes will lead to data corruption. Fully confirmed. The code that makes errors with die D does work well with die rev B.

**Comment by MJoergen** [comment]
Adding here a link to the errata : [https://issi.com/WW/pdf/appnotes/sram/AN66WX001_VariableMode_MinimumDataSize.pdf](https://issi.com/WW/pdf/appnotes/sram/AN66WX001_VariableMode_MinimumDataSize.pdf)

**Comment by AnttiLukats** [comment]
I modified the IP core to overcome the rev D errata and started "long time" testing, will post here the result.

**Comment by AnttiLukats** [comment]
It is running 4th day with ISSI die rev D, and no failures detected.  I leave it running.

**Comment by jorisvr** [comment]
Thanks for testing this!
I don't think I have ever seen 4 days without errors on TE0890. So _maybe_ those errors were specific to the older silicon revision and are solved in rev D.

Just for my understanding: are you now skipping short bursts, or running fixed latency, or extending short bursts to 2 cycles ?

**Comment by AnttiLukats** [comment]
I am extending the burst write for short bursts, so it is a real IP core fix! Status:
With rev A I did see ONE failure over the weekend
With rev D we are running for about a week and have no errors :(

Do you recall, when errors happened on TE0890 could it be that "all the error bundle" was within 1Kbyte of address range? If so then it is an indicator that there is a problem with REFRESH as that is done on the basis of 1K sectors.

What makes me wonder a bit is that with another HyperRAM IP core - OpenHBMC we do see rare errors on rev D too. But this test is also now running 4+ days and no errors. So PITA when you need error to come it ain't coming!

**Comment by jorisvr** [comment]
On the TE0890 I would see errors occurring in bursts. A burst would sometimes contain just one error, often a few or a hundred, but sometimes more than 10000 errors. So it was definitely not always true that the whole burst was confined to the same kbyte page.

The Poisson process is a nasty thing. At these low error rates you need extremely long test times to confirm _absence_ of errors to even a medium level of confidence.

**Comment by AnttiLukats** [comment]
yes, the memory test is running for a week now, but it does not yet confirm there are no errors coming. letting it run some more...

interesting that the error bundle can be larger than 1K block, hm...
